_pytest/python.py

```diff
@@ -1152,9 +1153,14 @@ def _idvalset(idx, valset, argnames, idfn, ids):
 def idmaker(argnames, argvalues, idfn=None, ids=None):
     ids = [_idvalset(valindex, valset, argnames, idfn, ids)
            for valindex, valset in enumerate(argvalues)]
-    if len(set(ids)) < len(ids):
-        # user may have provided a bad idfn which means the ids are not unique
-        ids = [str(i) + testid for i, testid in enumerate(ids)]
+    duplicates = [testid for testid in ids if ids.count(testid) > 1]
```

Comment:
A faster way to check for duplicates:

```python
has_duplicates = len(ids) != len(set(ids))
if has_duplicates:
```

While usually it won't matter, for a large list of test ids the difference might be significant:

```sh
$python -m timeit -s "ids = list(range(1000))" "[testid for testid in ids if ids.count(testid) > 1]"
100 loops, best of 3: 14.2 msec per loop

$python -m timeit -s "ids = list(range(1000))" "len(ids) != len(set(ids))"
10000 loops, best of 3: 22.2 usec per loop
```

If I increase the number of ids to `10000` I have to kill python after several seconds using the first snippet, while the second finishes in just ` 262 usec`. :wink: